PTV-1888

eapearson opened the pull request:
# PR for PTV-1888

The goal of PTV-1888 is to add development tools to the staging service, in preparation for a future PR (which will follow on the heels of this one) which fixes and improves support for file uploads of up to 5GB.

This PR contains the following changes:

- development tools, located in `./development`
- a `devcontainer` usable with Visual Studio Code located in `./devcontainer`
- Splitting Python dependencies into `requirements.txt` for the deployment runtime, and `requirements-dev.txt` containing additional dependencies for development and testing.
- Updates all dependencies - base image, base image dependencies, and Python dependencies
- Fixed a couple of unreliable tests
- Added markdownlint support, and applied markdownlint fixes to all markdown files, as I was working on documentation and need to to be "clean".
- Extracted and fixed API documentation, simplifying the top-level `README.md`.

## Development Tools

The development tools are designed to be run in a docker container, which may be a bespoke container defined `./development/Dockerfile` and controlled by a script `./development/scripts/run`, or may be the aforementioned `devcontainer`.

These tools include `mypy`, `black`, `isort`, and the `pytest`-based tests.

## `devcontainer`

Devcontainers are a relatively recent addition to the developer toolkit. Visual Studio Code (VSC) is already supported by this repo, so this can be seen as an addition to that support. The devcontainer has its own Dockerfile. For those unfamiliar with devcontainers, it is essentially a method for running VSC inside a container, and that container is configured to support the programming language and environment of choice. This is a very effective technique, as it replicates the precise system, system dependencies, python version, and python dependencies.

## Splitting Requirements

As the number of development support tools expands to cover current practices, the entanglement of runtime and development-time  dependencies becomes more serious. Therefore, the dependencies were simply split into "requirements.txt" and "requirements-dev.txt". The production image installs only "requirements.txt", and the devcontainer and development tools images also install the requirements-dev.txt.

## Update Dependencies

It is always good to take the opportunity to ensure that dependencies are up-to-date, to reduce the diff to dependencies in subsequence pull requests. In this case it was triggered by two other changes - the need to add new dependencies to support development tools, and the need to update the images due to a problem with the current Dockerfiles.

There is nothing else notable about the dependency updates.

## Dockerfile Update

I can briefly describe the Dockerfile issue. It appears that the Python base image dropped support for recent Python versions and the Debian "buster" releases. The effect of this was that the image would no longer build. So the Dockerfile was updated to use the base image `python:3.11.5-slim-bookworm`. This is a minor Python version bump from `3.11.4`,  a shift from Debian `buster` to `bookworm`, and an update of all the OS-level dependencies installed in the Dockerfile.

## Fixed Tests

There were two types of test fixes required.

First, the tests had used an small patch for running `hypothesis` in asynchronous tests. Some tests seemed to occasionally fail when tearing down the file system utility used to prepare a directory for tests to use. There seemed to be a race condition, and sometimes a file which was assumed to be deleted was not, and the operation to remove a directory used for test data would fail. I had assumed this was in the "custom" async adapter for the hypothesis library. It was during that investigation that I discovered that the custom adapter was no longer necessary, as hypothesis now works fine with asyncio-based tests. However, after this change, the test failures persisted, and I determined that there should not be any race conditions, as the async tests _should_ be sequenced (although there is always a doubt that concurrency will creep in). The next conjecture was that it was due to the interaction between the async client mock and the file utility. It ends up that the async client mock needs should be shut down before the file utility cleanup code, and thus the async client should be wrapped in the file utility . Applying this change to all such tests fixed the problem.

The second test problem was due to a time comparison for the modifiction type of file metadata. The test asserts that the when creates a file, the modification type should be earlier than any time measurement taken afterwards. However, sometimes this assertion failed! The reason? It is possible that the modification time obtained for a file may have a rough precision - in this case it was 1 second, although research showed that in windows it may be 2 seconds. In such a case, the modification time may be rounded to the nearest second, which may end up being the next second. Since the elapsed wall time between the file creation and the test code that generates the time comparison is so short, sub-second for sure, it is possible for the file creation time to be greater than the comparison time. This is not breaking the laws of sequential programming - it is just due to the modification time being rounded to the _next_ second.

The solution was to only look at the absolute difference between the two times, and ensure that it is less than 1 second. 

BTW this file modification precision issue was noticed only on my host machine - it did not replicate on GHA for example. This is probably due to the underling virtual machine used on a Rancher Desktop on macOS.

## Markdown Work

To support the new development tools, I added  additional documentation. Since I now relied on being able to edit markdown, I also added markdownlint support. This in turn made it convenient fix the markdown linting issues in all markdown files. There are not many, so it did not take much time. 

Markdown linting is not integrated into the development toolchain, just used environmentally in VSC. Support is in the form of a VSC plugin in the devocontainer configuration, and a `.markdownlint.yaml` configuration file.

## Extracted API Documentation

More time-consuming was extracting the API documention from the top-level README.md. I was editing the top level README in order to reflect the new development tools, and found that the length of the file was onerous to deal with - it  was over 1K lines. And the documentation format was difficult to read, and did not follow any kind of standard format. I hope the improvements I made will improve that situation, and I didn't quite complete it as some of the later code examples are somewhat difficult to follow. I did reformat them somewhat, but they are still cumbersome to read.

eapearson committed:
add a devcontainer [PTV-1888]

eapearson committed:
update Dockerfiles to use python 3.11.5,  bullseye, and fix os dependencies [PTV-1888]
python 3.11 no longer supports debian "buster".

eapearson committed:
add development tools [PTV-1888]

eapearson committed:
ignore jetbrains idea directory and temp dir [PTV-1888]
being nice to developers

eapearson committed:
split Python dependencies into service and development, and update them [PTV-1888]

eapearson committed:
fix flaky tests [PTV-1888]
they were broken for a few reasons: ordering of fileutil & app client, invalid time comparisons, and an async workaround for hypothesis is no longer necessary

eapearson committed:
fix local config [PTV-1888]

eapearson committed:
add docker plugin to devcontainer [PTV-1888]

eapearson committed:
put some work into docs [PTV-1888]
move api and development docs from README to separate, linked docs; add docs for dev tools and dockerfiles

eapearson committed:
Merge remote-tracking branch 'origin/develop' into PTV-1888

eapearson committed:
update release notes, update GHA test to use python 3.11.5

eapearson committed:
fix sonar warning [PTV-1888]

eapearson committed:
code quality gatekeeper changes [PTV-1888]

eapearson committed:
run tests in our container [PTV-1888]

eapearson committed:
run tests in our container [PTV-1888]

eapearson committed:
improve script invocation [PTV-1888]

eapearson committed:
somehow wasn't bookworm

eapearson committed:
don't run as regular user. [PTV-1888]

eapearson committed:
be more tolerant of different st_mtime implementations [PTV-1888]

eapearson committed:
erik, you are too tired [PTV-1888]

eapearson committed:
clarify test condition comment [PTV-1888]

eapearson committed:
add markdown lint, fix linting errors in all markdown files [PTV-1888]

eapearson committed:
remove old, unused travis config [PTV-1888]

eapearson committed:
top level scripts have moved to the scripts directory [PTV-1888]

eapearson committed:
fix typos [PTV-1888]

eapearson committed:
fix doc [PTV-1888]

eapearson committed:
adjust mtime test precision.

eapearson committed:
remove spaces (codacy) [PTV-1888]

eapearson commented:
Hi @bio-boris, I'm working through the code quality reports, and one big area is markdown linting issues. These are reported by codacy using "remark-lint". A few years ago they switched to "markdownlint" as the preferred markdown linting tool. Markdownlint is a much more popular tool than remark, as well. 
Would you mind switching to markdownlint in the codacy configuration? I can see the setting, but don't have admin access to be able to change it.
Thanks!

eapearson committed:
markdown formatting [PTV-1888]

eapearson commented:
Also, the sonarcloud error report for running a container as root. AFAIK this is a requirement for running a container in GHA (some people have tried to have a non-root user in the container match whatever user is used in GHA, but as far as I can tell this is not officially supported). 
I wonder if sonarcloud can be configured with an exception for this?

eapearson committed:
omit line for bandit/codacy not sonarcloud (codacy) [PTV-1888]

eapearson committed:
add openssh-client so can push to github inside devcontainer [PTV-1888]

eapearson committed:
okay, ignore in sonarcloud too [PTV-1888]

eapearson committed:
bullseye -> bookworm [PTV-1888]

bio-boris commented:
<img width="703" alt="image" src="https://github.com/kbase/staging_service/assets/1258634/c4a43431-8b0a-40ab-bbe9-b471415f52e3">
I have changed it

eapearson commented:
Thanks @bio-boris

bio-boris commented:
The SonarCloud is not a strict requirement for passing a PR review. It is for informational purposes to help us see if we did something wrong. I have marked it as Reviewed and now it is passing.

bio-boris commented:
I have triggered a re-run of codacy and it now passes. However, I noticed that you are using requirements.txt and requirements-dev Maybe we should instead get rid of requirements alltogether and switch to the PipFile? In the pipfile we can have both regular and dev dependencies

bio-boris commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+WORKDIR /app
+
+# Standard simplified python setup.
+COPY ./requirements.txt /app/requirements.txt
```
Maybe we should use pipfile instead?

eapearson replied:
The existing documentation says that the requirements.txt file is the one that is used, and that the Pipfile is there and (possibly) maintained for those who like to use it for development.  
Personally, I have never seen a backup package manager in a repo like this, and wish it weren't so. In some JS codebases we might have had npm and bower in the same repo, but that was a transition from an old package manager to a newer one.
I'd be ok with Pipfile, though I think if we switch the package manager Poetry would be a better choice than pipenv.
IMO It would be nice for KBase to support just 2 Python package managers - plain pip with requirements.txt (since it is so common) for legacy projects (and it is the one supported by kb-sdk), and Poetry for new repos.

bio-boris commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,39 @@
+	// "workspaceMount": "source=${localWorkspaceFolder},target=/workspace,type=bind,consistency=cached",
+	// The optional 'workspaceFolder' property is the path VS Code should open by default when
+	// connected. This is typically a file mount in .devcontainer/docker-compose.yml
+	"workspaceFolder": "/workspace",
```
maybe it would be better to call it `/workspaceFolder` to avoid confusion about the kbase workspace, just an idea

eapearson replied:
That is not my choice, it is the default configuration for VSC devcontainers.
Probably better not to change the default behavior, if possible. But certainly can be done!

bio-boris commented on .devcontainer/docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+    container_name: staging_server_devcontainer
+    dns: 8.8.8.8
+    volumes:
+      - ..:/workspace:cached
```
maybe it would be better to call it `/workspaceFolder` to avoid confusion about the kbase workspace, just an idea

eapearson replied:
That is not my choice, it is the default configuration for VSC devcontainers.

bio-boris commented on .devcontainer/docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+    # Don't worry, our main interaction with the container is through the 
+    # VSC terminal, which for a devcontainer opens a shell within the
+    # container.
+    command: /bin/sh -c "while sleep 1000; do :; done"
```
Is it possible to also open a debug server with this somehow and connect your IDE debugger into this with VSCODE? I believe it can be done in pycharm. Just wondering.

eapearson replied:
Yes, pretty sure, although I haven't tried.  When using a devcontainer you are actually running VSC in the container, so it is like running on whatever the platform is - in this case debian.

Unfortunately jetbrains is still catching up with devcontainers. They have initial support (and they have other similar things that are not as clean and smooth), but it is not very complete so can't be used for real work.

I typically run the service in development mode, and bang away at endpoints from an http tool, and then the front end, and use either a terminal in the devcontainer or the host tools to run tests.

In later work (the next PR), I have used these tools, esp. mypy, more extensively and will have a few changes. 

I find it effective to have cli tool support via a tool container and in the devcontainer, and to have those same tools available via VSC (or pycharm) plugins. It is a some work to set up, a pain in a codebase that is big and has not used them, and not insignificant maintenance, but I think that is the price for Python development these days.

(Don't get me started on all the extra work required to keep a Python codebase healthy; but I think ruff will be a good substitute for some of those tools and others not included here.)

bio-boris replied:
<img width="996" alt="image" src="https://github.com/kbase/staging_service/assets/1258634/7f745cd8-bd89-4378-9d92-ca9ffae54bb6">

Looks like env doesn't work on my machine

Upgraded my docker-compose still doesn't work

Docker Compose version 2.21.0

```
    environment:
      KB_DEPLOYMENT_CONFIG: ./deployment/conf/local.cfg
      FILE_LIFETIME: 90
```
might work

bio-boris commented on .github/workflows/run_tests.yml:
```diff
@@ -1,26 +1,12 @@
-        bash run_tests.sh
+    - name: Checkout the repo
+      uses: actions/checkout@v3
+
```
LGTM

bio-boris commented on .gitignore:
```diff
@@ -107,4 +107,10 @@ data/
 .DS_Store
```
LGTM

bio-boris commented on .markdownlint.yaml:
```diff
@@ -0,0 +1,10 @@
+MD013:
+  # Matches black line length for python code.
+  line_length: 88
```
Maybe we should consider having a larger line length? I believe most of us have large monitors nowadays. I kind of take it to the extreme and like 180 chars so I can keep string messages on the same line!

bio-boris commented on Dockerfile:
```diff
@@ -1,13 +1,13 @@
-FROM python:3.11.4-slim-buster
+FROM python:3.11.5-slim-bookworm
```
LGTM

bio-boris commented on README.md:
```diff
@@ -1,1237 +1,67 @@
+or from a locally-built image.
 
-## tests
+> TODO: provide barebones instructions here
```
TODO is present, will this be addressed in the next PR?

bio-boris replied:
@eapearson This issue is still here

bio-boris commented on README.md:
```diff
@@ -1,1237 +1,67 @@
-the debugger, set breakpoints and if you open the project in VSCode the debugger
-should be good to go. The provided configurations can run locally and run tests
-locally
+Development has [its own documentation](./docs/development/inde.md).
```
typo in the link

bio-boris replied:
@eapearson This issue is still here

bio-boris commented on README.md:
```diff
@@ -1,1237 +1,67 @@
+clearly one needs to have the ability to push up branches and create PRs.
 
-These are the currently supported upload app type IDs:
+Forks are not supported at this time.
```
In theory we are allowed to fork and then PR from inbox, but this is fine.

bio-boris commented on RELEASE_NOTES.md:
```diff
@@ -2,83 +2,102 @@
 
--  update to Python 3.11.4
--  run black on all service and test files, fix most linting complaints (pylint,
+- update to Python 3.11.5
```
LGTM

bio-boris commented on deployment/conf/local.cfg:
```diff
@@ -2,4 +2,5 @@
-CONCIERGE_PATH = /kbaseconcierge
\ No newline at end of file
+CONCIERGE_PATH = /kbaseconcierge
+FILE_EXTENSION_MAPPINGS = ./deployment/conf/supported_apps_w_extensions.json
```
+1

bio-boris commented on development/docker-compose-kbase-ui.yml:
```diff
@@ -0,0 +1,30 @@
+    environment:
+      - KB_DEPLOYMENT_CONFIG=/kb/deployment/conf/deployment.cfg
+      - FILE_LIFETIME=90
+      # - SAVE_STRATEGY=SAVE_STRATEGY_TEMP_THEN_COPY
```
LGTM, assuming this variables are coming in a different PR

bio-boris commented on development/scripts/README.md:
```diff
@@ -0,0 +1,35 @@
+# Scripts
```
+1

bio-boris commented on development/scripts/run:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
```
Should this be called run.sh?

bio-boris commented on development/tools/runner/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+
+# USER kbapp
+
+ENTRYPOINT [ "/app/development/tools/runner/entrypoint.sh" ]
```
I think we should use the identical dockerfile as the .devcontainer, and specify the entrypoint in the docker-compose.yaml instead. Having fewer dockerfiles to maintain might make things easier.

bio-boris replied:
We can also create a custom image and push it into the repo and use that as a base image.

bio-boris commented on docs/api.md:
```diff
@@ -0,0 +1,1425 @@
+
+- all paths should be specified treating the user's home directory as root
+
+- The url base, in an actualy deployment, will be:
```
typo

eapearson replied:
Fixed this and about a half dozen other typos.

bio-boris commented on docs/api.md:
```diff
@@ -0,0 +1,1425 @@
+
+## Test Service
+
+`GET /test-service`
```
This current serves as the /status endpoint. Maybe we can have 2 routes point to the same status function?

eapearson replied:
Yeah, we could, and should, but perhaps not in this PR.

bio-boris commented on docs/development/development-tools.md:
```diff
@@ -0,0 +1,119 @@
+### Running Tools
+
+All of the tools available for running from the host via docker are also directly
+available within the devcontainer. I still often run them from a host termianl anyway, as
```
typo

eapearson replied:
Yeah, fixing typos in that file. I didn't have the spellchecker running so those fingers do slip...

bio-boris commented on docs/development/maintaining-dockerfile.md:
```diff
@@ -0,0 +1,46 @@
+  - needs to copy all service files
+  - does not include development Python dependencies
+  - has the production entrypoint
+- `./.devcontainer/Dockerfile`
```
I think we could probably use one dockerfile and I think we could copy in the helper scripts and alternate endpoints in there as well. The entrypoint could also install required dev dependencies as well.

eapearson replied:
Ugh, I was hoping not to have to deal with multipurposing a single image.
What about using a Dockerfile include for the main setup? Basically the base could be from the base image up to the installation of Python service dependencies. This basically leaves the main service image to set up the entrypoint, and the development images to install the dev dependencies and have separate default entrypoints?

eapearson replied:
Well, I had read about Dockerfile INCLUDE, and not tried it ... I now have, and it isn't working (but perhaps that is just me, atm.) It also relies upon 3rd party extensions. The Docker team does not want to implement it, despite passionate and salty pleas: https://github.com/moby/moby/issues/735

bio-boris replied:
Alternative way is
```
# Base stage
FROM python:3.11 AS base
# Install shared dependencies, set up common environment...

# Production stage
FROM base AS production
# Copy service files, set up production-specific stuff...

# Development stage
FROM base AS development
# Set up dev-specific stuff...
```

docker build --target base -t myimage:base .
docker build --target production -t myimage:production .
docker build --target development -t myimage:development .


Could also use  --build-args as well

eapearson replied:
Nice! 
I'll have a crack at this, though I'd do it using docker compose files.

eapearson replied:
Okay, I've got this working locally.
Since I'm touching the Dockerfile with an eye to simplicity , I've been puzzled by some aspects of the codebase that aren't explained in any documents or comments.

A. I'm wondering why the entire repo is copied into /kb/module, rather than simply the components that are needed to run the service.
B. And I'm wondering why some bits ARE copied to /kb/deployment, when they are already present in /kb/module and could just be used from there (okay, I understand that one wants a more controlled deployment environment, but then why A)
C. And I'm wondering how the bits in the codebase are used that are not used in the service. E.g. scripts like scripts/prune_acls.py, refresh_token.py, or the entire top level import_specifications directory.
D. And there is deployment stuff like the top level hooks and build directories which seems defunct with GHA; unused? used in some other context?
E. Why is there an auth url in /deployment/conf/deployment.cfg that points ci? Presumably in prod the deployment config is volume mounted and KB_DEPLOYMENT_CONFIG is set to that path, and the one in the codebase is used just for local containers and possibly for a CI deployment.

eapearson replied:
Oh, and another question -
F. Wondering why the image installs htop and wget -- as conveniences for poking around a container? If they aren't used I'd remove them and only install in the dev target.

eapearson replied:
okay, @bio-boris , the Dockerfile and related changes are up now. I also took the opportunity to fold in some tool (black, isort, autoflake, flake8, mypy) improvements.

bio-boris replied:
```
3.200 Package wget is not available, but is referred to by another package.
3.200 This may mean that the package is missing, has been obsoleted, or
3.200 is only available from another source
3.200
3.202 E: Version '1.21.3-1+b2' for 'wget' was not found
------
WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
Dockerfile:26
--------------------
  25 |     # Install OS dependencies required by or nice-to-have in a development image
  26 | >>> RUN apt-get update && \
  27 | >>>     apt-get install -y --no-install-recommends htop=3.2.2-2 wget=1.21.3-1+b2 git=1:2.39.2-1.1 openssh-client=1:9.2p1-2 && \
  28 | >>>     apt-get clean && \
  29 | >>>     rm -rf /var/lib/apt/lists/*
  30 |     # Install Python dependencies require by development tools (cli tools and devcontainer)
--------------------
ERROR: failed to solve: process "/bin/sh -c apt-get update &&     apt-get install -y --no-install-recommends htop=3.2.2-2 wget=1.21.3-1+b2 git=1:2.39.2-1.1 openssh-client=1:9.2p1-2 &&     apt-get clean &&     rm -rf /var/lib/apt/lists/*" did not complete successfully: exit code: 100
docker build . --target dev
```

bio-boris replied:
Its very convenient to have wget and curl in all containers. They are very small and are very useful and also useful, especially when we get network connectivity issues.  I'd prefer not to remove that from the main image, as we can always install htop. In the distant, but not too distant future, we can use k8 debug containers, but until then it saves a lot of trouble by having them directly in the container. In fact, we could even use DockerSlim once we can use debug containers.

bio-boris replied:
Docs about some of those file are in our confluence

```
****************************
ACL And File Manager CronJob 
****************************

/root/bulk/acl_manager/ contains the required globus.cfg and python dependencies to run *prune_and_delete.py* 

The threshold to delete files and acls is defined as the constant THRESHOLD_DAYS, (currently set to 60)

If the modfied time of the user's directory is > THRESHOLD_DAYS, the ACL is marked as old, and deleted from globus, as well as the entire user's directory on */dtn/disk0/bulk/<username>*

The script should be run in the /root/bulk/acl_manager directory in order to access the *globus.cfg* and to consistently save logs in the same place.


See `here. <prune_and_delete.py>`_ 


****************************
ISSUES
****************************
* run /opt/kb-ftp-api/refresh_token.py to refresh token manually, otherwise this is run in a cronojb
* if it fails, uncomment the code inside to request a new token and rerun it
* if the staging service token fails, copy over the "REFRESH TOKENS" only into the appropriate values into the staging services globus.cfg file.
*

```

bio-boris commented on requirements.txt:
```diff
@@ -1,19 +1,14 @@
 frozendict==2.3.8
-globus-sdk==1.6.1
-hypothesis==6.81.1
+globus-sdk==3.28.0
```
I don't think we can upgrade globus

eapearson replied:
Ok, tests pass either way, I'll revert to the old version.

bio-boris reviewed (commented):
Nice changes and improvements! left some comments!

eapearson committed:
revert globus sdk version bump [PTV-1888]

eapearson committed:
fix typos [PTV-1899]

eapearson committed:
refactor Dockerfile to support dev and prod workflows [PTV-1888]
Uses multi-stage build to create a dev stage and production stage; removed other Dockerfiles, adjusted docker-compose and other things.

eapearson committed:
consolidate tool config into pyproject [PTV-1888]
includes aligning black to the previous line length of 100, aligns isort to black (otherwise they fight each other), set up mypy and autoflake, flake8 doesn't use pyproject

eapearson committed:
update python dependencies [PTV-1888]
Hold back globus_sdk

eapearson committed:
set pythonpath [PTV-1888]

eapearson committed:
apply black and isort after new configs [PTV-1888]

eapearson committed:
remove extra comma [PTV-1888]

bio-boris commented on development/tools/runner/entrypoint.sh:
```diff
@@ -0,0 +1,18 @@
+
+# Nice to set this up for all future python code being run.
+# We set the root of the repo as the python path. 
+export PYTHONPATH="${PWD}"
```
I am confused as to why this bash script is necessary instead of specifying the commands in the docker-compose.yaml file or the docker-compose binary.

ialarmedalien commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,40 @@
+				"tamasfe.even-better-toml",
+				"mhutchie.git-graph",
+				"ms-python.black-formatter",
+				"ms-python.flake8",
```
I suggest moving to [Ruff](https://docs.astral.sh/ruff/) -- it replaces most/all flake8 packages as well as adding its own rules, has an auto fix setting, and is much quicker than flake8. There's a VSCode integration for it: https://github.com/astral-sh/ruff-vscode

ialarmedalien replied:
On second thoughts, having finished going through the PR, leave Ruff for another PR. Not worth waiting to merge this one for a "nice to have".

ialarmedalien commented on development/scripts/run-dev-server.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
is there a reason why this doesn't use `/usr/bin/env bash` like the other scripts?

ialarmedalien commented on docs/development/development-tools.md:
```diff
@@ -0,0 +1,119 @@
+# Development Tools
+
+This document provides an overview of two different but related development approaches -
+devcontainers (via Visual Studio Code), and command line tools via a Docker container.
```
this is great! I've also been using vscode within running docker containers using VSCode's Docker extension. We should really have some sort of skill-share thing with anyone who does development for KBase as most people don't take advantage of what VSCode can do because it can be a PITA to set up.

ialarmedalien commented on tox.ini:
```diff
@@ -1,2 +1,5 @@
-max-line-length = 100
\ No newline at end of file
+# unfortunately, flake8 does not play with pyproject.tom. what a shame.
+# one day perhaps we'll switch to ruff.
```
Make that day today! It's very easy to migrate settings from flake8 to ruff, and also simple to switch out flake8 for ruff in the requirements file.
